src/libraries/System.Private.CoreLib/src/System/Array.cs

```diff
@@ -414,7 +422,260 @@ public static unsafe void Copy(Array sourceArray, int sourceIndex, Array destina
+        }
+
+        private static unsafe void CopyImpl(Array? sourceArray, int sourceIndex, Array? destinationArray, int destinationIndex, int length, bool reliable)
+        {
+            ArgumentNullException.ThrowIfNull(sourceArray);
+            ArgumentNullException.ThrowIfNull(destinationArray);
+
+            if (sourceArray.GetType() != destinationArray.GetType() && sourceArray.Rank != destinationArray.Rank)
+                throw new RankException(SR.Rank_MustMatch);
+
+            ArgumentOutOfRangeException.ThrowIfNegative(length);
+
+            int srcLB = sourceArray.GetLowerBound(0);
+            ArgumentOutOfRangeException.ThrowIfLessThan(sourceIndex, srcLB);
+            sourceIndex -= srcLB;
+            if ((sourceIndex < 0) || ((uint)(sourceIndex + length) > sourceArray.NativeLength))
+                throw new ArgumentException(SR.Arg_LongerThanSrcArray, nameof(sourceArray));
+
+            int dstLB = destinationArray.GetLowerBound(0);
+            ArgumentOutOfRangeException.ThrowIfLessThan(destinationIndex, dstLB);
+            destinationIndex -= dstLB;
+            if ((destinationIndex < 0) || ((uint)(destinationIndex + length) > destinationArray.NativeLength))
+                throw new ArgumentException(SR.Arg_LongerThanDestArray, nameof(destinationArray));
+
+            ArrayAssignType assignType;
+
+            if (sourceArray.GetType() == destinationArray.GetType()
+                || (assignType = CanAssignArrayType(sourceArray, destinationArray)) == ArrayAssignType.SimpleCopy)
+            {
+                MethodTable* pMT = RuntimeHelpers.GetMethodTable(sourceArray);
+
+                nuint elementSize = (nuint)pMT->ComponentSize;
+                nuint byteCount = (uint)length * elementSize;
+                ref byte src = ref Unsafe.AddByteOffset(ref MemoryMarshal.GetArrayDataReference(sourceArray), (uint)sourceIndex * elementSize);
+                ref byte dst = ref Unsafe.AddByteOffset(ref MemoryMarshal.GetArrayDataReference(destinationArray), (uint)destinationIndex * elementSize);
+
+                if (pMT->ContainsGCPointers)
+                    Buffer.BulkMoveWithWriteBarrier(ref dst, ref src, byteCount);
+                else
+                    SpanHelpers.Memmove(ref dst, ref src, byteCount);
+
+                // GC.KeepAlive(sourceArray) not required. pMT kept alive via sourceArray
+                return;
+            }
+
+            // If we were called from Array.ConstrainedCopy, ensure that the array copy
+            // is guaranteed to succeed.
+            if (reliable)
+                throw new ArrayTypeMismatchException(SR.ArrayTypeMismatch_ConstrainedCopy);
+
+            // Rare
+            CopySlow(sourceArray, sourceIndex, destinationArray, destinationIndex, length, assignType);
+        }
+
+        // Reliability-wise, this method will either possibly corrupt your
```

Comment:
This comment should be on CopyImpl. There is no reliable flag passed to this method.